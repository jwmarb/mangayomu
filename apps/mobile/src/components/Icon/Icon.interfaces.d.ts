import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';

import React from 'react';
import { TextProps } from '@components/Text/Text.interfaces';

export interface IconProps
  extends React.PropsWithChildren,
    Pick<TextProps, 'color' | 'variant' | 'bold' | 'italic' | 'underline'> {
  name: keyof typeof MaterialCommunityIconNames;
  size?: number;
}
