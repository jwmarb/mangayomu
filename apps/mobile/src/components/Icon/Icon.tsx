import styled, { css } from '@emotion/native';
import React from 'react';
import { IconProps, ImageIconProps } from './';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import { moderateScale, ScaledSheet } from 'react-native-size-matters';
import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import Animated from 'react-native-reanimated';
import { getOrUseCustomColor } from '@components/Text';
import { Image } from 'react-native';
import ImprovedImage from '@components/ImprovedImage';

const styles = ScaledSheet.create({
  imageIcon: {
    borderRadius: '4@ms',
  },
});

const AnimatedIcon = Animated.createAnimatedComponent(MaterialCommunityIcons);

const Icon: React.FC<IconProps | ImageIconProps> = (props) => {
  if (props.name == null) throw Error('name is required in Icon');
  if (props.type === 'font' && props.name in MaterialCommunityIconNames)
    return (
      <IconFromFont
        as={props.animated ? AnimatedIcon : undefined}
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        {...(props as unknown as any)}
      />
    );
  const { size = moderateScale(32) } = props;
  return (
    <ImprovedImage // ImprovedImage
      source={{ uri: props.name }}
      style={[styles.imageIcon, { width: size, height: size }]}
    />
  );
};

const IconFromFont = styled(MaterialCommunityIcons)<IconProps>`
  ${(props) => {
    const {
      theme,
      color = 'textPrimary',
      bold,
      italic,
      underline,
      variant = 'body',
      size,
    } = props;
    return css`
      ${variant !== 'inherit' && theme.typography[variant]};
      ${variant !== 'inherit' &&
      color !== 'inherit' &&
      `color: ${getOrUseCustomColor(theme, color)}`};
      ${size != null && 'font-size: ' + size + 'px'};
      ${bold && 'font-weight: bold'};
      ${italic && 'font-style: italic'};
      ${underline && 'text-decoration-line: underline'};
    `;
  }}
`;
export default Icon;
