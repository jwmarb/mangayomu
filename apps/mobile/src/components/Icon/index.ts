import Icon from './Icon';
export default Icon;
import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import React from 'react';
import { TextProps } from '@components/Text';

interface AbstractIconProps<T>
  extends React.PropsWithChildren,
    Pick<TextProps, 'color' | 'variant' | 'bold' | 'italic' | 'underline'> {
  type?: T;
}

export interface IconProps extends AbstractIconProps<'font'>, TextProps {
  name?: keyof typeof MaterialCommunityIconNames;
  size?: number;
  animated?: boolean;
}

export interface ImageIconProps extends AbstractIconProps<'image'> {
  name?: string;
  size?: number;
}
