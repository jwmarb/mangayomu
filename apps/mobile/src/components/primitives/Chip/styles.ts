import { ChipColors } from '@/components/primitives/types';
import { createStyles } from '@/utils/theme';

export const styles = (color: ChipColors | 'null') =>
  createStyles((theme) => ({
    chip_outlined: {
      borderColor:
        color === 'null' ? theme.palette.divider : theme.palette[color].main,
      borderRadius: theme.style.borderRadius / 3,
      borderWidth: 1.5,
      flexDirection: 'row',
      overflow: 'hidden',
    },
    chip_filled: {
      backgroundColor:
        color === 'null'
          ? theme.palette.divider
          : theme.palette[color][theme.mode === 'dark' ? 'dark' : 'light'],
      borderRadius: theme.style.borderRadius / 3,
      borderWidth: 1.5,
      flexDirection: 'row',
      overflow: 'hidden',
    },
    chipPressable: {
      flex: 0,
      flexDirection: 'row',
      paddingVertical: theme.style.size.m,
      paddingHorizontal: theme.style.size.l,
      alignItems: 'center',
      gap: theme.style.size.s,
    },
  }));
