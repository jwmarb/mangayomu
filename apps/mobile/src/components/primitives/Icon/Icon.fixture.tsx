import { SafeAreaView } from 'react-native-safe-area-context';
import { View } from 'react-native';
import { useSelect } from 'react-cosmos/client';
import MaterialCommunityIcons from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import Icon from '@/components/primitives/Icon';
import { createStyles } from '@/utils/theme';
import useStyles from '@/hooks/useStyles';
import Text from '@/components/primitives/Text';
import useModeSelect from '@/hooks/useModeSelect';
import { TEXT_COLORS, TextColors } from '@/components/primitives/types';

const styles = createStyles((theme) => ({
  container: {
    flex: 1,
    backgroundColor: theme.palette.background.default,
    justifyContent: 'center',
  },
  textContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    gap: theme.style.size.s,
  },
}));

export default function IconFixture() {
  const contrast = useModeSelect();
  const style = useStyles(styles, contrast);
  const [color] = useSelect('color', {
    options: TEXT_COLORS as unknown as string[],
    defaultValue: 'textPrimary',
  }) as unknown as [TextColors];
  const [name] = useSelect('icon-name', {
    options: Object.keys(MaterialCommunityIcons),
    defaultValue: 'ab-testing',
  }) as unknown as [keyof typeof MaterialCommunityIcons];
  return (
    <SafeAreaView style={style.container}>
      <View style={style.textContainer}>
        <Icon type="icon" name={name} size="small" color={color} />
        <Icon type="icon" name={name} size="medium" color={color} />
        <Icon type="icon" name={name} size="large" color={color} />
      </View>
      <View style={style.textContainer}>
        <Text variant="body2" color={color}>
          body2
        </Text>
        <Text color={color}>body1</Text>
        <Text variant="h3" color={color}>
          h3
        </Text>
      </View>
    </SafeAreaView>
  );
}
