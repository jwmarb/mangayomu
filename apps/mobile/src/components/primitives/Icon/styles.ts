import { StyleSheet, TextStyle } from 'react-native';
import { IconSizes } from '@/components/primitives/types';
import { IconProps } from '@/components/primitives/Icon/Icon';

export const sizes = (type: Pick<IconProps, 'type'>['type']) =>
  StyleSheet.create<Record<IconSizes, TextStyle>>({
    large:
      type === 'icon'
        ? {
            fontSize: 32,
          }
        : {
            width: 32,
            height: 32,
          },
    medium:
      type === 'icon'
        ? {
            fontSize: 24,
          }
        : {
            width: 24,
            height: 24,
          },
    small:
      type === 'icon'
        ? {
            fontSize: 16,
          }
        : {
            width: 16,
            height: 16,
          },
  });
