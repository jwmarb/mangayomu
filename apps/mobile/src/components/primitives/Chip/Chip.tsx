import React from 'react';
import { PressableProps, StyleProp, View, ViewStyle } from 'react-native';
import { styles } from '@/components/primitives/Chip/styles';
import Pressable from '@/components/primitives/Pressable';
import Text from '@/components/primitives/Text';
import {
  CHIP_COLORS,
  ChipColors,
  ChipVariants,
  TEXT_COLOR_TYPES,
} from '@/components/primitives/types';
import useContrast from '@/hooks/useContrast';
import useStyles from '@/hooks/useStyles';
import { IconProps } from '@/components/primitives/Icon';
import useTheme from '@/hooks/useTheme';
import { createThemedProps } from '@/utils/theme';
import useThemedProps from '@/hooks/useThemedProps';

export type ChipProps = {
  color?: ChipColors;
  contrast?: boolean;
  title: string;
  variant?: ChipVariants;
  icon?: React.ReactElement<IconProps>;
  style?: StyleProp<ViewStyle>;
} & Omit<PressableProps, 'style'>;

const composedStyles = [...CHIP_COLORS, 'null'].reduce((prev, curr) => {
  prev[curr] = styles(curr as ChipColors | 'null');
  return prev;
}, {} as Record<string, ReturnType<typeof styles>>);

const themedProps = [...CHIP_COLORS, 'null'].reduce((prev, curr) => {
  prev[curr] = TEXT_COLOR_TYPES.reduce((prev2, curr2) => {
    prev2[curr2] = createThemedProps((theme) => ({
      android_ripple: {
        color:
          curr === 'null'
            ? undefined
            : theme.palette[curr as ChipColors][curr2],
      },
    }));
    return prev2;
  }, {} as Record<string, ReturnType<typeof createThemedProps<{ android_ripple: { color: string | undefined } }>>>);
  return prev;
}, {} as Record<string, Record<string, ReturnType<typeof createThemedProps<{ android_ripple: { color: string | undefined } }>>>>);

function Chip(props: ChipProps) {
  const {
    title,
    color,
    contrast: contrastProp,
    variant = 'outlined',
    icon,
    style: styleProp,
    ...rest
  } = props;
  const theme = useTheme();
  const contrast = useContrast(contrastProp);
  const style = useStyles(
    composedStyles[color == null ? 'null' : color],
    contrast,
  );
  const colorType = !contrast
    ? theme.mode === 'light'
      ? 'dark'
      : 'light'
    : theme.mode === 'dark'
    ? 'dark'
    : 'light';
  const { android_ripple } = useThemedProps(
    themedProps[color == null ? 'null' : color].main,
    contrast,
  );
  const pressableStyle = [style.chipPressable, styleProp];

  return (
    <View style={style[`chip_${variant}` as keyof typeof style]}>
      <Pressable
        {...rest}
        android_ripple={android_ripple}
        style={pressableStyle}
      >
        {icon != null &&
          React.cloneElement(icon, { size: 'small', color, colorType })}
        <Text variant="chip" color={color} colorType={colorType}>
          {title}
        </Text>
      </Pressable>
    </View>
  );
}

export default React.memo(Chip);
