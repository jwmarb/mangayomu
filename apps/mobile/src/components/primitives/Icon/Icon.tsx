import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {
  Image,
  ImageStyle,
  StyleProp,
  TextProps,
  TextStyle,
} from 'react-native';
import React from 'react';
import Animated, { AnimatedProps } from 'react-native-reanimated';
import { sizes } from '@/components/primitives/Icon/styles';
import {
  IconSizes,
  TextColorTypes,
  TextColors,
} from '@/components/primitives/types';
import useStyles from '@/hooks/useStyles';
import useContrast from '@/hooks/useContrast';
import { composedColors } from '@/components/primitives/Text';

export type IconProps = BaseProps & (FontIconProps | ImageIconProps);

type BaseProps = Pick<TextProps, 'style'> & {
  color?: TextColors;
  colorType?: TextColorTypes;
  size?: 'large' | 'medium' | 'small' | number;
  contrast?: boolean;
  testID?: string;
};
type FontIconProps = {
  type: 'icon';
  name: keyof typeof MaterialCommunityIconNames;
};

type ImageIconProps = {
  type: 'image';
  uri: string;
};

export const composedSizes = (['icon', 'image'] as const).reduce(
  (prev, curr) => {
    prev[curr] = sizes(curr);
    return prev;
  },
  {} as Record<Pick<IconProps, 'type'>['type'], Record<IconSizes, TextStyle>>,
);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function Icon(props: IconProps, ref: React.ForwardedRef<any>) {
  const {
    color: colorProp = 'textPrimary',
    colorType = 'main',
    size: sizeProp = 'medium',
    contrast: contrastProp,
    style: styleProp,
    ...rest
  } = props;
  const contrast = useContrast(contrastProp);
  const size =
    typeof sizeProp === 'string'
      ? composedSizes[rest.type][sizeProp]
      : undefined;
  const color = useStyles(composedColors[colorType], contrast)[colorProp];
  const style =
    rest.type === 'icon' ? [size, color, styleProp] : [size, styleProp];
  switch (rest.type) {
    case 'icon':
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return (
        <MaterialCommunityIcons
          {...rest}
          ref={ref}
          name={rest.name}
          size={typeof sizeProp === 'number' ? sizeProp : undefined}
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          style={style as any}
        />
      );
    case 'image':
      return (
        <Image
          {...rest}
          ref={ref}
          src={rest.uri}
          style={style as StyleProp<ImageStyle>[]}
        />
      );
  }
}

const ForwardedIcon = React.forwardRef(Icon);

export const AnimatedIcon = Animated.createAnimatedComponent(
  ForwardedIcon,
) as React.FC<IconProps & AnimatedProps<IconProps>>;

export default ForwardedIcon as typeof Icon;
