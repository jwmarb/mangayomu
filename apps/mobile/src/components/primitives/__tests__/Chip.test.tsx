import { render, userEvent } from '@testing-library/react-native';
import Chip from '@/components/primitives/Chip';
import Icon from '@/components/primitives/Icon';

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

test('renders properly', () => {
  const tree = render(<Chip title="Hello World" />);
  expect(tree).toMatchSnapshot();
});

test('Is responsive', async () => {
  const onPress = jest.fn();
  const tree = render(<Chip testID="chip" onPress={onPress} title="Test" />);
  const user = userEvent.setup();

  await user.press(tree.getByTestId('chip'));

  expect(onPress).toHaveBeenCalled();

  jest.runOnlyPendingTimers();
});
