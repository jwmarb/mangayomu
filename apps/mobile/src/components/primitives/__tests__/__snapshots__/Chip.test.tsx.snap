// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<View
  style={
    {
      "borderColor": "rgba(0, 0, 0, 0.12)",
      "borderRadius": 10,
      "borderWidth": 1.5,
      "flexDirection": "row",
      "overflow": "hidden",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        [
          {
            "alignItems": "center",
            "flex": 0,
            "flexDirection": "row",
            "gap": 4,
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 13,
            "fontWeight": "normal",
            "letterSpacing": 0.15,
          },
          {
            "color": "rgba(0, 0, 0, 0.87)",
          },
          {
            "textAlign": "left",
          },
          [],
          [],
          undefined,
        ]
      }
    >
      Hello World
    </Text>
  </View>
</View>
`;
